File: src/lib/pricing.ts
// Rate-card pricing as the team uses it. Every price is a range with a usual
// ("standard") figure. Services priced by rows — employee tranches (Admin &
// PRO, Payroll & GOSI), workforce categories, accountancy rows, recruitment
// staff types, mobilization countries — carry those rows on the proposal line,
// each with its own price, and the deck's fee table lists exactly them
// (generator.rs / smartfill.rs). Pure.

import type { PricingService } from './constants';
import type { CommercialLine, LineRate } from './types';

export type PriceUnit = 'month' | 'one_time' | 'percent';

export interface PriceOption {
  label: string;
  min: number;
  standard: number;
  max: number;
  unit: PriceUnit;
}

/** How a service's proposal rows work. */
export type RowKind = 'tranche' | 'category' | 'row' | 'percent' | 'country';

export interface RowPreset {
  label: string;
  from: number | null;
  to: number | null;
  min: number;
  standard: number;
  max: number;
  percent: boolean;
  counts: boolean;
}

const round50 = (n: number) => Math.round(n / 50) * 50;
const standardOf = (min: number, max: number, standard?: number) => (standard != null && standard >= min && standard <= max ? standard : round50((min + max) / 2));

/** "1–5 employees" → 1..5; "25 employees and below" → 1..25. */
export function parseRange(label: string): { from: number; to: number } | null {
  const range = label.match(/(\d+)\s*[-–]\s*(\d+)/);
  if (range) return { from: Number(range[1]), to: Number(range[2]) };
  const below = label.match(/(\d+)\s*employees?\s*and\s*below/i);
  return below ? { from: 1, to: Number(below[1]) } : null;
}

export function trancheLabel(from: number | null | undefined, to: number | null | undefined): string {
  if (from != null && to != null) return `${from}–${to} employees`;
  if (to != null) return `Up to ${to} employees`;
  return from != null ? `${from}+ employees` : '';
}

export function rowKind(card: PricingService | null | undefined): RowKind | null {
  if (!card) return null;
  if (card.perCountry) return 'country';
  if (card.rows?.length) return card.rows.every((r) => r.percent) ? 'percent' : 'row';
  if (card.hasTranches && card.tranches?.length) return card.tranches.every((t) => parseRange(t.label)) ? 'tranche' : 'category';
  return null;
}

/** Rows a rate card suggests for a proposal, with or without commission. */
export function rowPresets(card: PricingService | null | undefined, commission = false): RowPreset[] {
  const kind = rowKind(card);
  if (!card || !kind) return [];
  if (kind === 'row' || kind === 'percent') {
    return (card.rows || []).map((r) => ({ label: r.label, from: null, to: null, min: r.min, standard: standardOf(r.min, r.max, r.standard), max: r.max, percent: !!r.percent, counts: !!r.counts }));
  }
  if (kind === 'country') return [];
  return (card.tranches || []).flatMap((t) => {
    const [min, max] = commission ? [t.commMin, t.commMax] : [t.noCommMin, t.noCommMax];
    if (min == null || max == null) return [];
    const range = parseRange(t.label);
    return [{ label: t.label, from: range?.from ?? null, to: range?.to ?? null, min, max, standard: standardOf(min, max), percent: false, counts: false }];
  });
}

/** The preset a row came from: same band limits, else the same label. */
export function presetFor(presets: RowPreset[], rate: LineRate): RowPreset | null {
  if (rate.to != null) {
    const band = presets.find((p) => p.to === rate.to && (p.from ?? null) === (rate.from ?? null));
    if (band) return band;
  }
  const label = rate.label.trim().toLowerCase();
  return presets.find((p) => p.label.toLowerCase() === label) || null;
}

const toRate = (p: RowPreset): LineRate => ({
  label: p.label,
  from: p.from,
  to: p.to,
  price: p.percent ? null : p.standard,
  percent: p.percent ? p.standard : null,
  counts: p.counts,
});

/**
 * The rows a new line starts with: three tranches (around the client's band
 * when the employee count is known), every workforce category and accountancy
 * row, the first recruitment staff type, one empty country.
 */
export function defaultRates(card: PricingService | null | undefined, commission = false, employeeCount?: number | null): LineRate[] {
  const kind = rowKind(card);
  const presets = rowPresets(card, commission);
  switch (kind) {
    case 'tranche': {
      const count = card?.showBands || 3;
      const at = employeeCount ? Math.max(0, presets.findIndex((p) => p.to != null && employeeCount <= p.to)) : 0;
      const start = Math.max(0, Math.min(at - 1, presets.length - count));
      return presets.slice(start, start + count).map(toRate);
    }
    case 'category':
    case 'row':
      return presets.map(toRate);
    case 'percent':
      return presets.slice(0, 1).map(toRate);
    case 'country':
      return [{ label: '', price: null, counts: false }];
    default:
      return [];
  }
}

/**
 * Monthly invoice for a number of employees across the tranches. Band 1 is a
 * minimum; above it, each extra employee costs (band price ÷ band upper limit),
 * added on top of the invoice at the previous band's limit.
 */
export function trancheInvoice(rates: LineRate[], employees: number): number | null {
  const bands = rates.filter((r) => r.to != null && r.to > 0 && r.price != null).sort((a, b) => a.to! - b.to!);
  if (!bands.length || employees <= 0) return null;
  let total = bands[0].price!;
  let prev = bands[0].to!;
  if (employees <= prev) return total;
  for (let i = 1; i < bands.length && prev < employees; i++) {
    const per = bands[i].price! / bands[i].to!;
    const upto = Math.min(employees, bands[i].to!);
    total += per * (upto - prev);
    prev = upto;
  }
  if (employees > prev) {
    const last = bands[bands.length - 1];
    total += (last.price! / last.to!) * (employees - prev);
  }
  return Math.round(total);
}

export interface LineValue {
  /** Monthly (or one-time) value used in totals; null when only rates are quoted. */
  value: number | null;
  /** How it was worked out, for the editor. */
  basis: 'employees' | 'minimum' | 'rows' | 'rates' | 'price';
}

export function lineValue(line: Pick<CommercialLine, 'rates' | 'employeeCount' | 'unitPrice' | 'quantity'>, card: PricingService | null | undefined): LineValue {
  const kind = rowKind(card);
  const rates = line.rates || [];
  if (!kind || !rates.length) return { value: line.unitPrice, basis: 'price' };
  if (kind === 'tranche') {
    if (line.employeeCount && line.employeeCount > 0) return { value: trancheInvoice(rates, line.employeeCount), basis: 'employees' };
    const lowest = rates.filter((r) => r.price != null).sort((a, b) => (a.to ?? Infinity) - (b.to ?? Infinity))[0];
    return { value: lowest?.price ?? null, basis: 'minimum' };
  }
  if (kind === 'row' || kind === 'country') {
    const counted = rates.filter((r) => r.counts && r.price != null);
    return counted.length ? { value: counted.reduce((a, r) => a + r.price!, 0), basis: 'rows' } : { value: null, basis: 'rates' };
  }
  return { value: null, basis: 'rates' };
}

// ── Single-range cards (Maintenance, Labor Law, Constitution, Liquidation…) ──

/** The standard price and range of a card priced as one figure. */
export function singleRange(card: PricingService | null | undefined, commission = false): PriceOption | null {
  if (!card || rowKind(card)) return null;
  const unit: PriceUnit = card.oneTime ? 'one_time' : 'month';
  if (card.percent) return { label: '', min: card.percent.min, standard: card.percent.standard, max: card.percent.max, unit: 'percent' };
  const [min, max] = commission && card.commMin != null && card.commMax != null ? [card.commMin, card.commMax] : [card.noCommMin, card.noCommMax];
  if (min == null || max == null) return null;
  return { label: '', min, max, standard: standardOf(min, max, commission ? undefined : card.standard), unit };
}

export type RangeCheck = 'below' | 'within' | 'above';

export function checkPrice(range: { min: number; max: number } | null, price: number | null | undefined): RangeCheck | null {
  if (!range || price == null) return null;
  return price < range.min ? 'below' : price > range.max ? 'above' : 'within';
}

File: src-tauri/src/pricing.rs
//! Rate-card pricing on the Rust side — the same rules as src/lib/pricing.ts:
//! how a service's proposal rows work (employee tranches, workforce
//! categories, accountancy rows, staff-type percentages, countries), and the
//! words used to match a row to a line in a template's fee table.

use serde::Deserialize;

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Tranche {
    pub label: String,
    pub no_comm_min: Option<f64>,
    pub no_comm_max: Option<f64>,
    pub comm_min: Option<f64>,
    pub comm_max: Option<f64>,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Percent {
    pub min: f64,
    pub standard: f64,
    pub max: f64,
    #[serde(default)]
    pub basis: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct PresetRow {
    pub label: String,
    #[serde(default)]
    pub min: f64,
    #[serde(default)]
    pub standard: f64,
    #[serde(default)]
    pub max: f64,
    #[serde(default)]
    pub percent: bool,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Card {
    #[serde(default)]
    pub has_tranches: bool,
    #[serde(default)]
    pub tranches: Vec<Tranche>,
    #[serde(default)]
    pub rows: Vec<PresetRow>,
    #[serde(default)]
    pub per_country: bool,
    #[serde(default)]
    pub percent: Option<Percent>,
    #[serde(default)]
    pub no_comm_min: Option<f64>,
    #[serde(default)]
    pub no_comm_max: Option<f64>,
    #[serde(default)]
    pub standard: Option<f64>,
    #[serde(default)]
    pub show_bands: Option<usize>,
}

impl Card {
    pub fn from_json(v: &serde_json::Value) -> Option<Card> {
        serde_json::from_value(v.clone()).ok()
    }

    /// The usual price of a card priced as one figure (Constitution 55,000).
    pub fn standard_price(&self) -> Option<f64> {
        match (self.no_comm_min, self.no_comm_max) {
            (Some(min), Some(max)) => Some(self.standard.filter(|s| *s >= min && *s <= max).unwrap_or(((min + max) / 2.0 / 50.0).round() * 50.0)),
            _ => self.standard,
        }
    }

    /// Labels of the rows this card suggests, to recognise its fee table rows.
    pub fn preset_labels(&self) -> Vec<String> {
        self.tranches.iter().map(|t| t.label.clone()).chain(self.rows.iter().map(|r| r.label.clone())).collect()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RowKind {
    Tranche,
    Category,
    Row,
    Percent,
    Country,
}

/// "1–5 employees" → (1, 5); "25 employees and below" → (1, 25).
pub fn parse_range(label: &str) -> Option<(i64, i64)> {
    static RANGE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    static BELOW: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    let range = RANGE.get_or_init(|| regex::Regex::new(r"(\d+)\s*[-–]\s*(\d+)").expect("regex"));
    if let Some(c) = range.captures(label) {
        return Some((c[1].parse().ok()?, c[2].parse().ok()?));
    }
    let below = BELOW.get_or_init(|| regex::Regex::new(r"(?i)(\d+)\s*employees?\s*and\s*below").expect("regex"));
    below.captures(label).and_then(|c| Some((1, c[1].parse().ok()?)))
}

pub fn tranche_label(from: Option<i64>, to: Option<i64>) -> String {
    match (from, to) {
        (Some(f), Some(t)) => format!("{f}–{t} employees"),
        (None, Some(t)) => format!("Up to {t} employees"),
        (Some(f), None) => format!("{f}+ employees"),
        _ => String::new(),
    }
}

pub fn row_kind(card: &Card) -> Option<RowKind> {
    if card.per_country {
        return Some(RowKind::Country);
    }
    if !card.rows.is_empty() {
        return Some(if card.rows.iter().all(|r| r.percent) { RowKind::Percent } else { RowKind::Row });
    }
    if card.has_tranches && !card.tranches.is_empty() {
        return Some(if card.tranches.iter().all(|t| parse_range(&t.label).is_some()) { RowKind::Tranche } else { RowKind::Category });
    }
    None
}

const IGNORED: &[&str] = &["and", "the", "of", "employee", "employees", "tranche", "mo"];

pub fn words(s: &str) -> Vec<String> {
    s.to_lowercase()
        .replace('&', " and ")
        .replace(['–', '—'], "-")
        .split(|c: char| !(c.is_ascii_alphanumeric() || c == '%' || c == '-'))
        .flat_map(|w| {
            // "Non-Nationalized" also reads as "non" + "nationalized".
            let parts: Vec<String> = w.split('-').filter(|p| !p.is_empty()).map(str::to_string).collect();
            std::iter::once(w.to_string()).chain(if parts.len() > 1 { parts } else { vec![] })
        })
        .filter(|w| !w.is_empty() && w != "-" && !IGNORED.contains(&w.as_str()))
        .collect()
}

/// Every word of `label` appears in `text` ("Nationalized (Engineers & Managers)"
/// matches "Professional Nationalized Employee (Engineers and Managers)").
pub fn names(label: &str, text: &str) -> bool {
    let lw = words(label);
    let tw = words(text);
    !lw.is_empty() && lw.iter().all(|w| tw.contains(w))
}

/// Like `names`, but a row for "Non-Nationalized" never counts as "Nationalized".
pub fn names_exactly(label: &str, text: &str) -> bool {
    if !names(label, text) {
        return false;
    }
    let lw = words(label);
    let tw = words(text);
    !(tw.contains(&"non".to_string()) && !lw.contains(&"non".to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reads_rate_cards_like_the_app() {
        let tr = |label: &str| Tranche { label: label.into(), no_comm_min: Some(1.0), no_comm_max: Some(2.0), ..Default::default() };
        let admin = Card { has_tranches: true, tranches: vec![tr("1–5 employees"), tr("6–15 employees")], ..Default::default() };
        let workforce = Card { has_tranches: true, tranches: vec![tr("Nationalized (Engineers & Managers)")], ..Default::default() };
        assert_eq!(row_kind(&admin), Some(RowKind::Tranche));
        assert_eq!(row_kind(&workforce), Some(RowKind::Category));
        assert_eq!(row_kind(&Card { per_country: true, ..Default::default() }), Some(RowKind::Country));
        assert_eq!(parse_range("Tranche 1 – PRO Services (25 Employees and Below)"), Some((1, 25)));
        assert_eq!(Card { no_comm_min: Some(45000.0), no_comm_max: Some(66000.0), standard: Some(55000.0), ..Default::default() }.standard_price(), Some(55000.0));
        assert!(names("Non-Nationalized (Unskilled)", "Professional Non-Nationalized Employee (Unskilled)"));
        assert!(!names_exactly("Nationalized (Engineers & Managers)", "Professional Non-Nationalized Employee (Engineers)"));
        assert!(names_exactly("Nationalized (Engineers & Managers)", "Professional Nationalized Employee (Engineers & Managers)"));
        assert!(names("Professional Staff", "Professional Staff Only"));
    }
}
